sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -0,0 +1,503 @@
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns></returns>
+        public async virtual Task<Response<UploadManifestResult>> UploadManifestAsync(OciManifest manifest, UploadManifestOptions options = default, CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNull(manifest, "manifest");
+
+            options ??= new UploadManifestOptions();
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(UploadManifest)}");
+            scope.Start();
+            try
+            {
+                var stream = SerializeManifest(manifest);
+                string tagOrDigest = options.Tag ?? OciBlobDescriptor.ComputeDigest(stream);
+                ResponseWithHeaders<ContainerRegistryCreateManifestHeaders> response = await _restClient.CreateManifestAsync(_repositoryName, tagOrDigest, manifest, ManifestMediaType.OciManifest.ToString(), cancellationToken).ConfigureAwait(false);
+                return Response.FromValue(new UploadManifestResult(response.Headers.DockerContentDigest), response.GetRawResponse());
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        private static Stream SerializeManifest(OciManifest manifest)
+        {
+            MemoryStream stream = new MemoryStream();
+            Utf8JsonWriter jsonWriter = new Utf8JsonWriter(stream);
+            ((IUtf8JsonSerializable)manifest).Write(jsonWriter);
+            jsonWriter.Flush();
+            return stream;
+        }
+
+        private static OciManifest DeserializeManifest(Stream stream)
+        {
+            using var document = JsonDocument.Parse(stream);
+            return OciManifest.DeserializeOciManifest(document.RootElement);
+        }
+
+        /// <summary>
+        /// Upload an artifact blob.
+        /// </summary>
+        /// <param name="stream">The stream containing the blob data.</param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns></returns>
+        public virtual Response<UploadBlobResult> UploadBlob(Stream stream, CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNull(stream, "stream");
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(UploadBlob)}");
+            scope.Start();
+            try
+            {
+                string digest = OciBlobDescriptor.ComputeDigest(stream);
+
+                ResponseWithHeaders<ContainerRegistryBlobStartUploadHeaders> startUploadResult =
+                    _blobRestClient.StartUpload(_repositoryName, cancellationToken);
+
+                ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult =
```

Comment:
Is there any size limit to this chunk upload? Just curious if we'd ever need to split this up into more than one chunk.

Reply:
Yes, we're not exposing chunk-by-chunk upload at the moment because of the level of complexity of the API and the fact that we don't have a current customer ask for it.  It's something we'll address if customers need it -- Docker changed their API from monolithic upload to these chunk APIs to enable their CLI to be able to resume a failed upload from a midpoint.

Reply:
interesting - it almost feels like an LRO